### ISSUE TITLE
Heap error from function with static number passed to byref param.

### ISSUE THREAD

**Issue by FortyTwoFortyTwo** [alliedmodders/sourcepawn#790]
This works fine in 1.10 and earlier versions of 1.11, but in `1.11.0.6832` I've been getting heap errors, which seems to be from byref param.

This code is what I currently have as the simplest method to create heap error, there might be even more simpler method than this to reproduce it.
```c++
public void OnPluginStart()
{
	for (int i = 0; i < 100; i++)
	{
		PrintToServer("%d", i);
		Funciton1();
	}
}

void Funciton1(int &ret = 0)
{
	//Create any dynamic array. Bigger array size produces less Funciton1 loops before heap error, and vice versa for more loops
	int[] output = new int[2048];
	
	//Need a bracket here, for loop can be used instead
	{
		if (Function2())
			return;	//This return is needed to create error, even if this seems unneeded
	}
}

bool Function2(int &ret = 0)	//Ref without being passed by variable
{
	return true;
}
```

This code expects to print a number from 0 to 99, but only manage to print 0 and 1 before getting heap error.
If a variable were to be passed to either `Funciton1` or `Function2` to fill byref parameter, heap error wouldn't appear.

I believe heap error also appears in forwards and callbacks that have byref param, as I had an error from sourcemod's `NormalSHook` callback that have several byref params.

**Comment by asherkin** [comment]
You're using quite an old compiler version, this looks like a bug that was fixed earlier this year.

**Comment by FortyTwoFortyTwo** [comment]
Still produces heap error with SM `1.11.0.6875` compiler

**Comment by JoinedSenses** [comment]
Off topic, but gh has highlighting for sp if you wanna use that instead of c++

\`\`\`sourcepawn
stock int f(int x) { return x; }
\`\`\`
```sourcepawn
stock int f(int x) { return x; }
```

**Comment by asherkin** [comment]
Bisected to cd3275790e330be8769439aaef3afc39b6979732 / #688 (unsurprisingly).

spshell-friendly test case:
```sourcepawn
#include <shell>

public void main()
{
        for (int i = 0; i < 100; i++)
        {
                printnum(i);
                Funciton1();
        }
}

void Funciton1(int &ret = 0)
{
        //Create any dynamic array. Bigger array size produces less Funciton1 loops before heap error, and vice versa for more loops
        int[] output = new int[2048];

        //Need a bracket here, for loop can be used instead
        {
                if (Function2())
                        return; //This return is needed to create error, even if this seems unneeded
        }
}

bool Function2(int &ret = 0)    //Ref without being passed by variable
{
        return true;
}
```

**Comment by asherkin** [comment]
I'm no expert but it looks like the heap tracking around `DefaultArgExpr` is all correct, I'm guessing the issue here is with the containing heap scope.

**Comment by dvander** [comment]
Thanks for the reduced test case, and indeed the bug is not in DefaultargExpr. RETN does not interact well with heap scopes. I ran out of time today to fix, but this should block 1.11.

**Comment by dvander** [comment]
This has been fixed on master.
